### PR TITLE
Feature editable text element

### PR DESCRIPTION
**Pull request by mdaxtman**
@rgerstenberger 

Please review. Code comments are there for things that look out of ordinary for clarification.

Will handle text input:
screen shot 2016-07-07 at 2 41 24 pm

Line Breaks:
screen shot 2016-07-07 at 2 41 39 pm

WordWrapping on resize:
screen shot 2016-07-07 at 2 41 47 pm

### CODE REVIEW

File: app/components/canvas/element-types/text-element/index.js [hunk 1]
```diff
@@ -457,28 +560,38 @@ export default class TextElement extends Component {
...
-                    {...props}
-                    style={{ ...elementStyle, ...computedResizeStyles }}
-                  >
-                    {children}
-                  </ComponentClass>
-                {currentlySelected &&
+                <TextContentEditor
+                  ref={component => {
+                    const element = ReactDOM.findDOMNode(component);
+                    if (element && element.getAttribute("contenteditable")) {
```
Comment (rgerstenberger) [inline app/components/canvas/element-types/text-element/index.js]: Can you explain this ref?

File: app/components/canvas/element-types/text-element/index.js [hunk 2]
```diff
@@ -356,24 +347,133 @@ export default class TextElement extends Component {
...
+
+  createUpdateElementChildren = () => {
+    const {
+      currentState,
+      currentSlideIndex,
+      currentElementIndex
+    } = this.context.store;
+
+    const currentElementChildren =
+      currentState.slides[currentSlideIndex].children[currentElementIndex].children;
```
Comment (rgerstenberger) [inline app/components/canvas/element-types/text-element/index.js]: You can just grab `currentElement` from the store.

File: app/components/canvas/element-types/text-element/index.js (hunk 1 above)
Comment (mdaxtman) [inline app/components/canvas/element-types/text-element/index.js]: that's a ref to can access to the `contenteditable` div for text editing. There's a ternary within that component that will one `div` or another. I did this to make sure that `this.editable` is only assigned to the div with the `contenteditable` attribute, and not the non-editable div that will be returned when not editing. I can change this condition to be `if (editing)` if that sounds better.

File: app/components/canvas/element-types/text-element/index.js [hunk 3]
```diff
@@ -356,24 +347,133 @@ export default class TextElement extends Component {
...
+        this.editable.removeEventListener("blur", updateElementChildren);
+        this.editable.removeEventListener("input", this.handleInput);
+        this.editable.blur();
+
+        const { currentContent } = this.state;
+
+        if (typeof currentContent === "string" && currentContent !== currentElementChildren) {
+          this.context.store.updateChildren(
+            trimEnd(currentContent, "\n"),
+            currentSlideIndex,
```
Comment (rgerstenberger) [inline app/components/canvas/element-types/text-element/index.js]: I don't think we need to pass around `currentSlideIndex` or `currentElementIndex` the store already knows which element is selected in `currentElement`

File: app/components/canvas/element-types/text-element/index.js (hunk 1 above)
Comment (rgerstenberger) [inline app/components/canvas/element-types/text-element/index.js]: Yeah, that sounds better.

File: app/components/canvas/element-types/text-element/index.js (hunk 2 above)
Comment (mdaxtman) [inline app/components/canvas/element-types/text-element/index.js]: See line notes for line 445.

File: app/components/canvas/element-types/text-element/index.js (hunk 3 above)
Comment (mdaxtman) [inline app/components/canvas/element-types/text-element/index.js]: Commenting per our conversation:

This is necessary to capture these integers as closure variables before the blur event occurs, as store can and will update `currentSlideIndex` and `currentElementIndex` before the blur event happens. This is essentially a way to keep track of the `currentElement` and its values when a text edit starts.

**Comment by rgerstenberger** [comment]
This LGTM but I'd like @kenwheeler to review as well before merging.

File: app/index.js [hunk 4]
```diff
@@ -5,6 +5,25 @@ import routes from "./routes";
...
 
+/**
+ * redefinition of console.error is done because
+ * app/components/canvas/element-types/text-element/text-content-editor.js
+ * uses attribute, "contentEditable", which alows for editing
+ * of DOM elements by user input. This can result in updates
+ * occuring on the DOM outside of the scope of react, and
+ * results in warnings that flood the console. This IIFE
+ * ignores those warnings
+ */
```
Comment (kenwheeler) [inline app/index.js]: What warning does this throw?

File: app/components/canvas/element-types/text-element/text-content-editor.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      content
+    } = this.props;
+
+    return isEditing ?
+      (<div
+        contentEditable="true"
+        ref={component => {this.editable = component;}}
+        {...componentProps}
+        className={`${contentClass} ${editorClass}`}
+        style={Object.assign(style, { whiteSpace: "pre-wrap" })}
```
Comment (kenwheeler) [inline app/components/canvas/element-types/text-element/text-content-editor.js]: This could be spread right?

**Comment by kenwheeler** [comment]
Looks solid in my book

File: app/index.js (hunk 4 above)
Comment (mdaxtman) [inline app/index.js]: here's a link to a closed issue that was filed with react in regards to this warning.

https://github.com/facebook/react/issues/2477

File: app/components/canvas/element-types/text-element/text-content-editor.js (hunk 5 above)
Comment (mdaxtman) [inline app/components/canvas/element-types/text-element/text-content-editor.js]: Updated.

**Comment by mdaxtman** [comment]
@rgerstenberger @kenwheeler I'll merge if everything is a-ok. 👍 ?

File: app/components/canvas/element-types/text-element/text-content-editor.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      content
+    } = this.props;
+
+    return isEditing ?
+      (<div
+        contentEditable="true"
+        ref={component => {this.editable = component;}}
+        {...componentProps}
+        className={`${contentClass} ${editorClass}`}
+        style={{ ...style, ...{ whiteSpace: "pre-wrap" } }}
```
Comment (rgerstenberger) [inline app/components/canvas/element-types/text-element/text-content-editor.js]: This can just be `style={{ ...style, whiteSpace: "pre-wrap" }}`

**Comment by rgerstenberger** [comment]
One minor nit, fix then merge :)

**Pull request merged**